modules/apps/27-interchain-accounts/host/types/query.go

```diff
@@ -0,0 +1,76 @@
+package types
+
+import (
+	"fmt"
+	"strings"
+
+	gogoproto "github.com/cosmos/gogoproto/proto"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+
+	queryv1 "cosmossdk.io/api/cosmos/query/v1"
+	errorsmod "cosmossdk.io/errors"
+
+	ibcerrors "github.com/cosmos/ibc-go/v8/modules/core/errors"
+)
+
+// IsModuleSafeQuery checks whether the method with the given grpcServicePath has the
+// `(cosmos.query.v1.module_query_safe) = true` proto annotation.
+//
+// For example, `/cosmos.bank.v1beta1.Query/Balance` is module safe, but
+// `/cosmos.reflection.v1.ReflectionService/FileDescriptors` is not.
+func IsModuleQuerySafe(grpcServicePath string) (bool, error) {
+	methodPath, err := toMethodPath(grpcServicePath)
+	if err != nil {
+		return false, err
+	}
+
+	protoFiles, err := gogoproto.MergedRegistry()
+	if err != nil {
+		// This should never happen
+		panic(err)
+	}
+	if protoFiles == nil {
+		protoFiles = protoregistry.GlobalFiles
+	}
+
+	fullName := protoreflect.FullName(methodPath)
+	if !fullName.IsValid() {
+		return false, errorsmod.Wrap(ibcerrors.ErrInvalidRequest, fmt.Sprintf("invalid method path: %s", methodPath))
```

Comment:
These should be turned into logs and simply return false instead since protoFiles may be non-deterministic and change between patch releases